### PR TITLE
Add separate dev file system

### PR DESCRIPTION
**Pull request by vtikoo**
Supported entries: /dev/urandom, /dev/zero, /dev/null and /dev/fd.

Signed-off-by: Vikas Tikoo <vikasamar@gmail.com>

### CODE REVIEW

File: kernel/devfs.c [hunk 1]
```diff
@@ -0,0 +1,2502 @@
...
+**==============================================================================
+**
+** inode_t
+**
+**==============================================================================
+*/
+
+#define INODE_MAGIC 0xcdfbdd61258a4c9d
+
+struct inode
```
[the remark this answers is not in the document]
Comment (vtikoo) [reply kernel/devfs.c:76]: The inode here has some extra callback fields - `read_vcallback` and `write_vcallback`. These are introduced in this PR to implement unbounded files like /dev/urandom, /dev/zero, /dev/null.

**Comment by vtikoo** [comment]
@jxyang  Agree that the code duplication isn't ideal. There are two ways I can think of improving this -
- devfs has a private ramfs handle, and forwards file operations to ramfs.
- extend ramfs to include changes to read/write callbacks. Then devfs can be like procfs, an instance of ramfs mounted at `/dev`.

cc: @mikbras

**Comment by vtikoo** [comment]
Some details on read/write callbacks:

The existing virtual file implementation adds a `v_callback` field to the `inode` and a `v_buf` field to `myst_file_t`.
`v_buf` is populated at open(), and released at close(). The `v_buf` is also populated(and released) during stat() calls to calculate file size.
This implementation is used for static /proc entries like `/proc/meminfo` and `/proc/self/` symlink.

This however is insufficient for unbounded files like `/dev/urandom` or `/dev/zero`. To support these files, `read_callback` and `write_callback` were added to the `inode` structure. A virtual file can either have a read/write callback or open callback.
Reads and writes are directly writing into the user's provided buffer.

**Comment by jxyang** [comment]
Maybe define a `Virtual Ramfs` that we can unify procfs, devfs, and future virtual file systems. But you don't have do that in this PR.

**Comment by vtikoo** [comment]
```
Maybe define a Virtual Ramfs that we can unify procfs, devfs, and future virtual file systems. 
```
@jxyang I did give this some thought - separating virtual files related code as a layer on top of ramfs. But couldn't think of a clean way to do it.

PR #323 tries to extend virtual files in ramfs to be able to support /dev entries as well.

**Comment by vtikoo** [comment]
Abandoned in favor of #323.

**Pull request closed without merging**